pgex/examples/collision_shapes/main.py

```diff
@@ -0,0 +1,335 @@
+    """ Check collision between polygon and rectangle. """
+
+    # IMPORTANT: A better approach would be clipping the polygon and testing
+    #            against the rect. But for now we are just treating the rect
+    #            as another polygon.
+
+    rect_center = pygame.Vector2(*rect.center)
+
+    rect_vertices = [
+        pygame.Vector2(*rect.topleft) - rect_center,
+        pygame.Vector2(*rect.topright) - rect_center,
+        pygame.Vector2(*rect.bottomright) - rect_center,
+        pygame.Vector2(*rect.bottomleft) - rect_center
+    ]
+
+    rect_poly = Polygon(rect_center, rect_vertices)
+
+    return polygon_x_polygon(polygon, rect_poly)
+
+def polygon_x_polygon(polygon_a: Polygon, polygon_b: Polygon) -> bool:
+    """ Check collision between two polygons. """
+
+    vertices_a = polygon_a.local_to_world()
+    vertices_b = polygon_b.local_to_world()
+    na = len(vertices_a)
+    nb = len(vertices_b)
+
+    # Check for polygon A's edges
+    for i in range(na):
+        va = vertices_a[i]
+        vb = vertices_a[(i + 1) % na]
+
+        edge = vb - va
+        axis = edge.normalize()
+        axis = pygame.Vector2(-axis.y, axis.x)
+
+        min_a, max_a = project_polygon(polygon_a, axis)
+        min_b, max_b = project_polygon(polygon_b, axis)
+
+        if min_a >= max_b or min_b >= max_a:
+            return False
+    
+    # Check for polygon B's edges
+    for i in range(nb):
+        va = vertices_b[i]
+        vb = vertices_b[(i + 1) % nb]
+
+        edge = vb - va
+        axis = edge.normalize()
+        axis = pygame.Vector2(-axis.y, axis.x)
+
+        min_a, max_a = project_polygon(polygon_a, axis)
+        min_b, max_b = project_polygon(polygon_b, axis)
+
+        if min_a >= max_b or min_b >= max_a:
+            return False
+        
+    return True
+
+
+async def main() -> None:
+    pygame.init()
+    window = pygame.display.set_mode((450, 140))
+    pygame.display.set_caption("PGEX Collision Shapes Example")
+    clock = pygame.time.Clock()
+    
+    # Prepare scene
+
+    circles = []
+
+    circles.append( Circle(pygame.Vector2(50, 50), 30) )
+    circles.append( Circle(pygame.Vector2(90, 90), 20) )
+    circles.append( Circle(pygame.Vector2(330, 90), 15) )
+
+    rects = []
+
+    rects.append( pygame.FRect(113, 25, 80, 50) )
+    rects.append( pygame.FRect(180, 50, 30, 65) )
+
+    polygons = []
+
+    pentagon_vertices = [
+        pygame.Vector2(25, 34), pygame.Vector2(40, -13), pygame.Vector2(0, -43),
+        pygame.Vector2(-40, -13), pygame.Vector2(-25, 34)
+    ]
+
+    triangle_vertices = [
+        pygame.Vector2(27, 27),
+        pygame.Vector2(-37, 10),
+        pygame.Vector2(10, -37)
+    ]
+
+    polygons.append( Polygon(pygame.Vector2(270, 50), pentagon_vertices) )
+    polygons.append( Polygon(pygame.Vector2(370, 50), triangle_vertices) )
+
+    # This list will be used to check collision between all shapes
+    # Check the note in line 177
```

Comment:
Huh, I strongly advice against this kind of comments.

You should never have comments that complicate to maintain.
The code can evolve a lot through time, and people (or you) adding code around line 177 will 100% **not** update this comment accordingly.
Proof is this comment is already obsolete: the note is on line 179.